cedar-policy-core/src/parser/cst_to_ast.rs

```diff
@@ -1473,53 +1473,13 @@ impl Node<Option<cst::Mult>> {
                     }
                 }
             }
-            // split all the operands into constantints and nonconstantints.
-            // also, remove the opcodes -- from here on we assume they're all
-            // `Times`, having checked above that this is the case
-            let (constantints, nonconstantints): (Vec<ast::Expr>, Vec<ast::Expr>) =
-                std::iter::once(first)
-                    .chain(more.into_iter().map(|(_, e)| e))
-                    .partition(|e| {
-                        matches!(e.expr_kind(), ast::ExprKind::Lit(ast::Literal::Long(_)))
-                    });
-            let constantints = constantints
-                .into_iter()
-                .map(|e| match e.expr_kind() {
-                    ast::ExprKind::Lit(ast::Literal::Long(i)) => *i,
-                    // PANIC SAFETY Checked the match above via the call to `partition`
-                    #[allow(clippy::unreachable)]
-                    _ => unreachable!(
-                        "checked it matched ast::ExprKind::Lit(ast::Literal::Long(_)) above"
-                    ),
-                })
-                .collect::<Vec<Integer>>();
-            if nonconstantints.len() > 1 {
-                // at most one of the operands in `a * b * c * d * ...` can be a nonconstantint
-                errs.push(self.to_ast_err(ToASTErrorKind::NonConstantMultiplication));
-                None
-            } else if nonconstantints.is_empty() {
-                // PANIC SAFETY If nonconstantints is empty then constantints must have at least one value
-                #[allow(clippy::indexing_slicing)]
-                Some(ExprOrSpecial::Expr {
-                    expr: construct_expr_mul(
-                        construct_expr_num(constantints[0], self.loc.clone()),
-                        constantints[1..].iter().copied(),
-                        &self.loc,
-                    ),
-                    loc: self.loc.clone(),
-                })
-            } else {
-                // PANIC SAFETY Checked above that `nonconstantints` has at least one element
-                #[allow(clippy::expect_used)]
-                let nonconstantint: ast::Expr = nonconstantints
-                    .into_iter()
-                    .next()
-                    .expect("already checked that it's not empty");
-                Some(ExprOrSpecial::Expr {
-                    expr: construct_expr_mul(nonconstantint, constantints, &self.loc),
-                    loc: self.loc.clone(),
-                })
-            }
+            // remove the opcodes -- from here on we assume they're all `Times`,
+            // having checked above that this is the case
+            let more = more.into_iter().map(|(_, e)| e);
```

Comment:
We could do this via a `filter_map` instead of having the invariant

Reply:
Refactored.  Possible small behavior change in the case where `first.into_expr` throws an error and there was a `/` or `%` -- we used to prefer the `first.into_expr` error and now we'll prefer the `/` or `%` error.  But I think that's acceptable.